### PR TITLE
View all products incl. name, price and sku

### LINKED ISSUE #6: View all products name, price, sku

### PR DESCRIPTION
**Pull request by craigjbass**
Fixes #6

### CODE REVIEW

File: src/main/kotlin/uk/co/craigbass/pratura/domain/Product.kt [hunk 1]
```diff
@@ -2,4 +2,4 @@ package uk.co.craigbass.pratura.domain
 
 import java.math.BigDecimal
 
-class Product(var sku: String, var price: BigDecimal)
\ No newline at end of file
+class Product(var sku: String, var price: BigDecimal, val name: String)
```
Comment (tomverran) [inline src/main/kotlin/uk/co/craigbass/pratura/domain/Product.kt]: I dunno what idiomatic Kotlin looks nice but in Scala it is nice to wrap stringy values that aren't arbitrary user-facing text into a single valued data class (this is about Sku)

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt [hunk 2]
```diff
@@ -18,7 +18,15 @@ class ViewBasket(private val basketItemsGateway: BasketItemsRetriever,
     return PresentableBasket(lineItems(), "£${getBasketValue()}")
   }
 
-  private fun lineItems() = items!!.map { item -> PresentableLineItem(item.quantity, item.sku) }
+  private fun lineItems(): List<PresentableLineItem> {
+    return items!!.map { item ->
```
Comment (tomverran) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt]: What do these double !!s do? They look dangerous

File: src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt [hunk 3]
```diff
@@ -9,23 +9,33 @@ import uk.co.craigbass.pratura.testdouble.InMemoryProductGateway
 import uk.co.craigbass.pratura.usecase.AddProduct
 
 class AddProductSpec : Spek({
-
   val productGateway = InMemoryProductGateway()
-
+  val useCase = memoized { AddProduct(productGateway) }
```
Comment (tomverran) [inline src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt]: Again probably I'm missing something here but what is the effect of memoising a val?

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package uk.co.craigbass.pratura.usecase
+
+import uk.co.craigbass.pratura.boundary.ViewAllProducts
+import uk.co.craigbass.pratura.boundary.ViewAllProducts.PresentableProduct
+import java.math.BigDecimal
+
+class ViewAllProducts(val productRetriever: ProductRetriever) : ViewAllProducts {
+  override fun execute(request: Unit): List<PresentableProduct> {
+    return productRetriever.all().map { product ->
+      PresentableProduct(product.name, "£${product.price.toCurrencyString()}", product.sku)
```
Comment (tomverran) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt]: This being production code the hard coded £ seems a bit dubious, or is localisation for the future?

File: src/main/kotlin/uk/co/craigbass/pratura/domain/Product.kt (hunk 1 above)
Comment (tomverran) [inline src/main/kotlin/uk/co/craigbass/pratura/domain/Product.kt]: Looks like* not looks nice

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt (hunk 4 above)
Comment (tomverran) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt]: Also huh I just noticed your unit parameter, I guess that is to fit in with your must have a request value interface?

File: src/main/kotlin/uk/co/craigbass/pratura/domain/Product.kt (hunk 1 above)
Comment (tomverran) [inline src/main/kotlin/uk/co/craigbass/pratura/domain/Product.kt]: Also why are these things vars?

Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/domain/Product.kt]: Should be `val` good catch! `val` = immutable, `var` = mutable

Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/domain/Product.kt]: >I dunno what idiomatic Kotlin looks nice but in Scala it is nice to wrap stringy values that aren't arbitrary user-facing text into a single valued data class (this is about Sku)

Yeah you can use type aliases in Kotlin, but not played around much with this

File: src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package uk.co.craigbass.pratura.boundary
+
+import com.madetech.clean.usecase.SynchronousUseCase
+import uk.co.craigbass.pratura.boundary.ViewAllProducts.PresentableProduct
+
+interface ViewAllProducts : SynchronousUseCase<Unit, List<PresentableProduct>> {
+  data class PresentableProduct(val name: String, val price: String, val sku: String)
+}
```
Comment (patrickkerrigan) [inline src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt]: A little petty, but the inconsistent file endings keep catching my eye

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt (hunk 2 above)
Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt]: yeah this will throw an exception if `items` is null. It's fine, I have a test suite :smile:

File: src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt (hunk 3 above)
Comment (craigjbass) [inline src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt]: In `Spek` test framework, a memoized value is read-only and cleared between each test "it" block

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt (hunk 2 above)
Comment (tomverran) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt]: Given that lists have a natural empty value why make the list nullable?

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt (hunk 4 above)
Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt]: Yeah Unit largely means Nothing in Kotlin. it's because all concrete "SynchronousUseCase" implementations need a request param... but here we have nothing in our request.

Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt]: I created a handy extension method to remove the tedium of `execute(Unit)`

File: src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt (hunk 5 above)
Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt]: You're worse than rubocop

Comment (patrickkerrigan) [inline src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt]: I've been spending too much time setting up linting on Jenkins :laughing:

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt (hunk 2 above)
Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt]: The list isn't set with a value until `execute` is called

File: src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt (hunk 3 above)
Comment (tomverran) [inline src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt]: Ah ok 👍

File: src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt (hunk 5 above)
Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt]: Not a massive fan of linting, gotta say

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt (hunk 2 above)
Comment (tomverran) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt]: Oh hmm sharing values between functions with nullable fields rather than parameters is a bit eyebrow raising but the scope is indeed small

Also wouldn't you still want to show individual prices against your PresentableLineItems?

File: src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt (hunk 5 above)
Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt]: Fixed

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt (hunk 2 above)
Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt]: Yeah got an issue for that open #16

File: src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt (hunk 5 above)
Comment (tomverran) [inline src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt]: Yeah I am unkeen on linting in compiled languages but even I switched on the IntelliJ auto end of file newline

Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/boundary/ViewAllProducts.kt]: Yeah I've turned it on.. lols

File: src/main/kotlin/com/madetech/clean/boundary/AsynchronousUseCaseExecutor.kt [hunk 6]
```diff
@@ -5,4 +5,4 @@ import kotlin.reflect.KClass
 
 interface AsynchronousUseCaseExecutor {
   fun <U : AsynchronousUseCase<R, P>, R, P> executeUseCase(useCase: KClass<U>, request: R, presenter: P)
```
Comment (paulmorrishill) [inline src/main/kotlin/com/madetech/clean/boundary/AsynchronousUseCaseExecutor.kt]: don't like generic executors at all

File: src/main/kotlin/com/madetech/clean/usecase/SynchronousUseCase.kt [hunk 7]
```diff
@@ -2,4 +2,6 @@ package com.madetech.clean.usecase
 
 interface SynchronousUseCase<in REQUEST, out RESPONSE> {
```
Comment (paulmorrishill) [inline src/main/kotlin/com/madetech/clean/usecase/SynchronousUseCase.kt]: Do not like generic use case interfaces at all

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ProductRetriever.kt [hunk 8]
```diff
@@ -4,4 +4,4 @@ import uk.co.craigbass.pratura.domain.Product
 
 interface ProductRetriever {
   fun all(): List<Product>
```
Comment (paulmorrishill) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ProductRetriever.kt]: methods should be verbs or verb phrases. e.g. ```getAll()```

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt (hunk 4 above)
Comment (paulmorrishill) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt]: Don't use bigdecimal for currency, create a Money data type which has the value and the currency.

Comment (paulmorrishill) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt]: Dispatch to the currency of the money type for the symbol to render.

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt [hunk 9]
```diff
@@ -18,7 +18,15 @@ class ViewBasket(private val basketItemsGateway: BasketItemsRetriever,
...
+      PresentableLineItem(
+        quantity = item.quantity,
+        sku = item.sku,
+        name = getProductFor(item)?.name ?: ""
```
Comment (paulmorrishill) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt]: ugly line

File: src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt [hunk 10]
```diff
@@ -32,7 +36,10 @@ class BasketSpec : Spek({
     beforeEachTest {
       pratura().executeUseCase(
```
Comment (paulmorrishill) [inline src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt]: why are you testing through the use case executor, this is obvious boilerplate.

File: src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt [hunk 11]
```diff
@@ -73,5 +87,39 @@ class BasketSpec : Spek({
         basketContents().basketValue.shouldEqual("£10.00")
```
Comment (paulmorrishill) [inline src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt]: You should dispatch this to a currency formatter which takes the Money data type and outputs a string. Dependency inject different formatter based on culture of user using system.

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt (hunk 9 above)
Comment (tomverran) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt]: Is it valid for items to not have products? Better to just explode if not

File: src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt [hunk 12]
```diff
@@ -9,24 +9,32 @@ import uk.co.craigbass.pratura.testdouble.InMemoryProductGateway
...
   beforeEachTest {
-    AddProduct(productGateway).execute(Request("sku:1029317", "1.00".toDecimal()))
+    useCase().execute(Request(
```
Comment (paulmorrishill) [inline src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt]: useCase should be called something like ```addProduct().execute(``` more clear

File: src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt [hunk 13]
```diff
@@ -9,24 +9,32 @@ import uk.co.craigbass.pratura.testdouble.InMemoryProductGateway
...
+      sku = "sku:1029317",
+      price = "6.99".toDecimal(),
+      name = "Pack of 4 AA Batteries"
+    ))
   }
 
+  val savedProducts = memoized { productGateway.all() }
+  val firstSavedProduct = memoized { savedProducts().first() }
+
   it("should have saved one product") {
```
Comment (paulmorrishill) [inline src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt]: I would combine these assertion blocks into one e.g. "It should have saved the correct product information" then 3 assertion lines. Having all this boiler plate for 1 assertion line is equivalent to commenting every line. If your code is readable then there should be no harm in combining them. It is one logical assertion after all.

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt (hunk 4 above)
Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewAllProducts.kt]: Yeah, will implement the concept of currencies as a separate thing. Think there is a whole bunch of complexity here

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt (hunk 9 above)
Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ViewBasket.kt]: Yeah I'm glossing over a lot of edge cases atm

File: src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt (hunk 10 above)
Comment (craigjbass) [inline src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt]: This is an acceptance test, it should test through the boundary

File: src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt (hunk 11 above)
Comment (craigjbass) [inline src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt]: Currencies are complex, glossing over this for now

File: src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt (hunk 12 above)
Comment (craigjbass) [inline src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt]: this is the AddProduct spec. We will only test one useCase here.

File: src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt (hunk 13 above)
Comment (craigjbass) [inline src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt]: Yeah maybe

File: src/main/kotlin/uk/co/craigbass/pratura/usecase/ProductRetriever.kt (hunk 8 above)
Comment (craigjbass) [inline src/main/kotlin/uk/co/craigbass/pratura/usecase/ProductRetriever.kt]: Yeah agree

File: src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt (hunk 10 above)
Comment (paulmorrishill) [inline src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt]: where are the unit tests?

File: src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt (hunk 11 above)
Comment (paulmorrishill) [inline src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt]: simple thing to implement makes life easier

File: src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt (hunk 12 above)
Comment (paulmorrishill) [inline src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt]: yeah doesn't change the fact it's more intuitive.

File: src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt (hunk 13 above)
Comment (paulmorrishill) [inline src/test/kotlin/uk/co/craigbass/pratura/unit/AddProductSpec.kt]: Yeah definitely.

File: src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt (hunk 10 above)
Comment (craigjbass) [inline src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt]: For what part?

File: src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt (hunk 11 above)
Comment (craigjbass) [inline src/test/kotlin/uk/co/craigbass/pratura/acceptance/BasketSpec.kt]: Dunno - want to defer this decision
